eax/src/lib.rs

```diff
@@ -0,0 +1,299 @@
+//! methods accept any type that impls the [`aead::Buffer`] trait which
+//! contains the plaintext for encryption or ciphertext for decryption.
+//!
+//! Note that if you enable the `heapless` feature of this crate,
+//! you will receive an impl of [`aead::Buffer`] for `heapless::Vec`
+//! (re-exported from the [`aead`] crate as [`aead::heapless::Vec`]),
+//! which can then be passed as the `buffer` parameter to the in-place encrypt
+//! and decrypt methods:
+//!
+//! ```
+//! # #[cfg(feature = "heapless")]
+//! # {
+//! use eax::EaxAes256; // Or `EaxAes128`
+//! use eax::aead::{AeadInPlace, NewAead, generic_array::GenericArray};
+//! use eax::aead::heapless::{Vec, consts::U128};
+//!
+//! let key = GenericArray::from_slice(b"an example very very secret key.");
+//! let cipher = EaxAes256::new(key);
+//!
+//! let nonce = GenericArray::from_slice(b"my unique nonces"); // 128-bits; unique per message
+//!
+//! let mut buffer: Vec<u8, U128> = Vec::new();
+//! buffer.extend_from_slice(b"plaintext message");
+//!
+//! // Encrypt `buffer` in-place, replacing the plaintext contents with ciphertext
+//! cipher.encrypt_in_place(nonce, b"", &mut buffer).expect("encryption failure!");
+//!
+//! // `buffer` now contains the message ciphertext
+//! assert_ne!(&buffer, b"plaintext message");
+//!
+//! // Decrypt `buffer` in-place, replacing its ciphertext context with the original plaintext
+//! cipher.decrypt_in_place(nonce, b"", &mut buffer).expect("decryption failure!");
+//! assert_eq!(&buffer, b"plaintext message");
+//! # }
+//! ```
+//!
+//! [1]: https://en.wikipedia.org/wiki/Authenticated_encryption
+
+#![no_std]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![doc(html_logo_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo_small.png")]
+#![deny(unsafe_code)]
+#![warn(missing_docs, rust_2018_idioms)]
+
+#[cfg(test)]
+#[macro_use]
+extern crate std;
+
+pub use aead::{self, AeadInPlace, Error, NewAead};
+
+#[cfg(feature = "aes")]
+pub use aes;
+
+use block_cipher::{
+    consts::{U0, U16},
+    generic_array::functional::FunctionalSequence,
+    generic_array::{ArrayLength, GenericArray},
+    Block, BlockCipher, Key, NewBlockCipher,
+};
+use cmac::crypto_mac::NewMac;
+use cmac::{Cmac, Mac};
+use ctr::stream_cipher::{FromBlockCipher, SyncStreamCipher};
+
+#[cfg(feature = "aes")]
+use aes::{Aes128, Aes256};
+
+// TODO Max values?
+/// Maximum length of associated data
+pub const A_MAX: u64 = 1 << 36;
+
+/// Maximum length of plaintext
+pub const P_MAX: u64 = 1 << 36;
+
+/// Maximum length of ciphertext
+pub const C_MAX: u64 = (1 << 36) + 16;
+
+/// EAX tags
+pub type Tag = GenericArray<u8, U16>;
+
+/// EAX with a 128-bit key and 96-bit nonce
+#[cfg(feature = "aes")]
+#[cfg_attr(docsrs, doc(cfg(feature = "aes")))]
+pub type EaxAes128 = Eax<Aes128>;
+
+/// EAX with a 256-bit key and 96-bit nonce
+#[cfg(feature = "aes")]
+#[cfg_attr(docsrs, doc(cfg(feature = "aes")))]
+pub type EaxAes256 = Eax<Aes256>;
```

Comment:
I think typically the mode comes second after the cipher and these should be `AesEax128` and `AesEax256`.